packages/runtime/src/ctx.ts

```diff
@@ -0,0 +1,370 @@
+      freezeTarget,
+      this.state.accounts.get(freezeTarget)
+    );
+    acc.setFreezeState(assetId, freezeState);
+  }
+
+  /**
+   * https://developer.algorand.org/docs/features/asa/#revoking-an-asset
+   * Revoking an asset for an account removes a specific number of the asset
+   * from the revoke target account.
+   * @param recipient asset receiver address
+   * @param assetId asset index
+   * @param revocationTarget revoke target account
+   * @param amount amount of assets
+   */
+  revokeAsset (
+    recipient: string, assetID: number,
+    revocationTarget: string, amount: number
+  ): void {
+    // Transfer assets
+    const fromAssetHolding = this.getAssetHolding(assetID, revocationTarget);
+    const toAssetHolding = this.getAssetHolding(assetID, recipient);
+
+    if (fromAssetHolding.amount - amount < 0) {
+      throw new RuntimeError(RUNTIME_ERRORS.TRANSACTION.INSUFFICIENT_ACCOUNT_ASSETS, {
+        amount: amount,
+        address: revocationTarget
+      });
+    }
+    fromAssetHolding.amount -= amount;
+    toAssetHolding.amount += amount;
+  }
+
+  /**
+   * https://developer.algorand.org/docs/features/asa/#destroying-an-asset
+   * Destroy asset
+   * @param sender sender's address
+   * @param assetId asset index
+   * @param payFlags transaction parameters
+   */
+  destroyAsset (assetId: number): void {
+    const creatorAcc = this.getAssetAccount(assetId);
+    // destroy asset from creator's account
+    creatorAcc.destroyAsset(assetId);
+    // delete asset holdings from all accounts
+    this.state.accounts.forEach((value, key) => {
+      value.assets.delete(assetId);
+    });
+  }
+
+  /**
+   * Delete application from account's state and global state
+   * @param appId Application Index
+   */
+  deleteApp (appId: number): void {
+    if (!this.state.globalApps.has(appId)) {
+      throw new RuntimeError(RUNTIME_ERRORS.GENERAL.APP_NOT_FOUND, { appId: appId, line: 'unknown' });
+    }
+    const accountAddr = this.runtime.assertAddressDefined(this.state.globalApps.get(appId));
+    if (accountAddr === undefined) {
+      throw new RuntimeError(RUNTIME_ERRORS.GENERAL.ACCOUNT_DOES_NOT_EXIST);
+    }
+    const account = this.runtime.assertAccountDefined(
+      accountAddr, this.state.accounts.get(accountAddr)
+    );
+
+    account.deleteApp(appId);
+    this.state.globalApps.delete(appId);
+  }
+
+  /**
+   * Closes application from account's state
+   * @param sender Sender address
+   * @param appId application index
+   */
+  closeApp (sender: AccountAddress, appId: number): void {
+    const fromAccount = this.getAccount(sender);
+    // https://developer.algorand.org/docs/reference/cli/goal/app/closeout/#search-overlay
+    this.runtime.assertAppDefined(appId, this.getApp(appId));
+    fromAccount.closeApp(appId); // remove app from local state
+  }
+
+  /**
+   * Process transactions in ctx
+   * - Runs TEAL code if associated with transaction
+   * - Executes the transaction on ctx
```

Comment:
Please also add the explanation (if not added somewhere). That we're doing  this because if any one tx in group fails, then it does not affect `runtime.store`, otherwise we just update `store` with `ctx` (if all transactions are executed successfully).